external/fv3fit/fv3fit/reservoir/readout.py

```diff
@@ -20,36 +27,115 @@ def square_even_terms(v: np.ndarray, axis=1) -> np.ndarray:
 
-    def predict(self, input):
-        if len(input.shape) == 1:
-            input = input.reshape(1, -1)
+    def predict(self, input: np.ndarray):
         if self.square_half_hidden_state:
-            input = square_even_terms(input, axis=1)
-        return self.linear_regressor.predict(input)
+            input = square_even_terms(input, axis=0)
+        return np.dot(self.coefficients, input) + self.intercepts
 
     def dumps(self) -> bytes:
         components = {
-            "linear_regressor": self.linear_regressor,
-            "square_half_hidden_state": self.square_half_hidden_state,
+            "hyperparameters": dataclasses.asdict(self.hyperparameters),
+            "coefficients": self.coefficients,
+            "intercepts": self.intercepts,
         }
         f = io.BytesIO()
         joblib.dump(components, f)
         return f.getvalue()
+
+
+class CombinedReservoirComputingReadout:
+    """Combines readout layers of multiple reservoir computing models
+    into a block diagonal readout layer, which can be used to predict over
+    the combined domain of those models.
+    Prediction is done on vector of concatenated reservoir states.
+
+    linear_regressors: sequence of sklearn Ridge regressors, each corresponding
+        to a subdomain
+    square_half_hidden_state: if True, square even terms in the reservoir state
+        before it is used as input to the regressor's .fit and .predict methods
+        This option was found to be important for skillful predictions in
+        Wikner+2020 (https://doi.org/10.1063/5.0005541)
+    """
+
+    _READOUT_NAME = "readout.bin"
+
+    def __init__(self, readouts: Sequence[ReservoirComputingReadout]):
+        self._combine_readouts(readouts)
+
+    def _combine_readouts(self, readouts: Sequence[ReservoirComputingReadout]):
+        coefs, intercepts, square_state_settings = [], [], []
+        for readout in readouts:
+            coefs.append(readout.coefficients)
+            intercepts.append(readout.intercepts)
+            square_state_settings.append(readout.square_half_hidden_state)
+
+        # Merge the coefficient arrays of individual readouts into single
+        # block diagonal matrix
+        self.coefficients = scipy.sparse.block_diag(coefs)
+
+        # Concatenate the intercepts of individual readouts into single array
+        self.intercepts = np.concatenate(intercepts)
+
+        if len(np.unique(square_state_settings)) != 1:
+            raise ValueError(
+                "All readouts must have the same setting for square_half_hidden_state."
+            )
+        self.square_half_hidden_state = square_state_settings[0]
+
+    def predict(self, input: np.ndarray):
+        if self.square_half_hidden_state:
+            input = square_even_terms(input, axis=0)
+        return self.coefficients * input + self.intercepts
+
+    @classmethod
+    def load(cls, paths: Sequence[str]) -> "CombinedReservoirComputingReadout":
```

Comment:
Is it going to be feasible to configure/pass this a list of paths? Or should the class itself be responsible for organizing and loading sub-paths from one path? I notice there isn't a `save` method on this.

Reply:
Good point, configuring a list of >10s of paths would be annoying. Changed to list and read from sorted subdirs in `path`. I didn't include a dump method on this class because I figured the saving of each readout to the specified subdir could be handled by the training function.

Reply:
The one thing to keep in mind with that is the responsibility for knowing how sub-models are stored at a low level now exists in two places - your code that writes the models, and this code that reads them. Ideally you'd be able to remove that low level detail from the writing code and keep it on this class. One possible way to do something like this would be to use the same `get_combined_reservoir_paths(path, n_reservoirs)` helper function in both places.

Reply:
That's a good idea. I just realized this part needs to be redone before merging- right now I don't think it's possible to use `ReservoirComputingModel.dump` for multiple models and save the readouts to the directory structure assumed here. Combined with the other suggestion to not dump readout/reservoir to the same path, the new assumed directory structure when reading and writing multiple subdomain models is
```
dir
|- model_0
   |- readout
       |- weights, etc.
   |- reservoir
       |- weights, etc.
|- model_1
    |- ...
```
`CombinedReservoirComputingReadout.load` will take `dir` as the path arg and search for directories ending in numbers, then open their readout subdirs.